[#1405] Add TraitAdvancement & TraitConfig

arbron opened the pull request:
Adds the `TraitAdvancement` type and the configuration application.

<img width="607" alt="Screenshot 2023-10-25 at 12 26 57" src="https://github.com/foundryvtt/dnd5e/assets/19979839/f041b17d-25f9-46e5-9516-e1e915a6bac3">

Fyorl reviewed (changes requested):
Thanks for all the hard work on this, Jeff. This is a very complex configuration workflow that you've managed to represent in a very compact way.

I had a little bit of trouble really understanding what was going on and how it worked at first, but I got the hang of it eventually and it makes sense. I think since we view configuring custom advancements more of an advanced tool, we can tolerate a higher level of complexity in the UI.

I had some suggestions for ways we might be able to make the dialog a little more intuitive initially though. The radio buttons threw me off at first since they seemed to imply I was making some configuration choice, but actually I was just toggling between which set I was configuring at a time. One option that came to mind was doing away with the in-place editing entirely. We would essentially move the right-hand side panel inside its own dialog that is spawned when you choose to configure a given set by clicking on its cog icon.

![image](https://github.com/foundryvtt/dnd5e/assets/207433/2641b3d7-4412-4f58-a1b0-11ac7bbc1ff4)

You can add and remove sets as before, then click the cog to configure which items appear in each set.

This would allow us to move the 'count' configuration inside this new dialog too. I also struggled a bit with the count as it was part of the same fieldset, and wasn't immediately clear that I was configuring a count for the selected set rather than some sort of count for the total number of options or something.

I appreciate it would be a significant rewrite so I'm not sure how palatable a suggestion it is, but I thought I would throw it out there in case it resonated strongly with you.

If we don't want to go with that suggestion, another idea I had was moving the count configuration 'inline' with the set description. So each one would always read at least '1 from X, Y, Z', where the '1' is actually inside its own inline number input. That would make the count configuration a lot clearer to me at least, though I appreciate it complicates the form handling somewhat.

Similarly, if we do not want to go with the separate dialogs suggestion, then I think my main issue was with the radio buttons. If we had some alternative method of highlighting the 'selected' set that wasn't a radio button (some bespoke CSS styling might do), and somehow connecting that to the fields that are used to configure the selected set (the checkboxes on the right, and the 'count' configuration), then I think it would be an improvement too.

arbron commented:
@Fyorl I've made a pass at changing the styling of the sheet:

<img width="652" alt="Screenshot 2023-10-31 at 17 11 34" src="https://github.com/foundryvtt/dnd5e/assets/19979839/82a38fb9-8279-460c-b9fc-21ecb0ee7c48">

I moved Count to the right column with the trait choices, replaced the radio buttons with a cog and arrow for selected. I've also made the selected option bold for better distinction.

I swapped out the minus icon for a trash can because the minus icon felt a bit too visually similar to the arrow.

I increase the gap between the two columns to give them some visual separation.

Fyorl reviewed (approved):
I think the changes address most of the major points of friction I had, thanks for making those. Especially the tools list gets very long and unwieldy so I hope we could have some better way of displaying these trait options eventually. That applies to not only this dialog but also the trait dialogs on the character sheet.

I've used checkbox grids to good effect in other designs, but they only really allow for one 'layer' of category nesting. That should be fine for the core dnd5e traits that we support, but it won't generalise to a full tree unfortunately.

![image](https://github.com/foundryvtt/dnd5e/assets/207433/aa16e321-3af5-4260-88dc-7ac68d863d29)

Either way, I think the design you've implemented currently is good and we can move ahead with it and iterate on it further later if we feel it's necessary.
